src/wayland/shell/xdg/mod.rs

```diff
@@ -175,88 +463,254 @@ impl PositionerState {
+            geometry.y -= geometry.height;
+        } else if self.gravity_has_edge(xdg_positioner::Gravity::Bottom) {
+            geometry.y -= geometry.height / 2;
+        }
+
+        if self.gravity_has_edge(xdg_positioner::Gravity::Left) {
+            geometry.x -= geometry.width;
+        } else if self.gravity_has_edge(xdg_positioner::Gravity::Right) {
+            geometry.x -= geometry.width / 2;
+        }
+
+        geometry
+    }
 }
 
 /// State of a regular toplevel surface
-#[derive(Debug)]
+#[derive(Debug, PartialEq)]
 pub struct ToplevelState {
-    /// Parent of this surface
-    ///
-    /// If this surface has a parent, it should be hidden
-    /// or displayed, brought up at the same time as it.
-    pub parent: Option<wl_surface::WlSurface>,
-    /// Title of this shell surface
-    pub title: String,
-    /// App id for this shell surface
-    ///
-    /// This identifier can be used to group surface together
-    /// as being several instance of the same app. This can
-    /// also be used as the D-Bus name for the app.
-    pub app_id: String,
-    /// Minimum size requested for this surface
-    ///
-    /// A value of 0 on an axis means this axis is not constrained
-    pub min_size: (i32, i32),
-    /// Maximum size requested for this surface
-    ///
-    /// A value of 0 on an axis means this axis is not constrained
-    pub max_size: (i32, i32),
+    /// The suggested size of the surface
+    pub size: Option<(i32, i32)>,
+
+    /// The states for this surface
+    pub states: ToplevelStates,
+
+    /// The output for a fullscreen display
+    pub fullscreen_output: Option<wl_output::WlOutput>,
+}
+
+impl Default for ToplevelState {
+    fn default() -> Self {
+        ToplevelState {
+            fullscreen_output: None,
+            states: Default::default(),
+            size: None,
+        }
+    }
 }
 
 impl Clone for ToplevelState {
     fn clone(&self) -> ToplevelState {
         ToplevelState {
-            parent: self.parent.as_ref().cloned(),
-            title: self.title.clone(),
-            app_id: self.app_id.clone(),
-            min_size: self.min_size,
-            max_size: self.max_size,
+            fullscreen_output: self.fullscreen_output.clone(),
+            states: self.states.clone(),
+            size: self.size,
         }
     }
 }
 
-/// The pending state of a popup surface
-#[derive(Debug)]
-pub struct PopupState {
-    /// Parent of this popup surface
-    pub parent: Option<wl_surface::WlSurface>,
-    /// The positioner specifying how this tooltip should
-    /// be placed relative to its parent.
-    pub positioner: PositionerState,
+/// Container holding the states for a `XdgToplevel`
+///
+/// This container will prevent the `XdgToplevel` from
+/// having the same `xdg_toplevel::State` multiple times
+/// and simplifies setting and un-setting a particularly
+/// `xdg_toplevel::State`
+#[derive(Debug, Clone, PartialEq)]
+pub struct ToplevelStates {
```

Comment:
Maybe we can find a better name for this, it's very easy to confuse with `ToplevelState`... Maybe something like `StateSet` ?

Reply:
Renamed to ToplevelStateSet